### DISCUSSION TITLE
Architecture Design for K3s High Availability with an External DB vs Embedded DB

### DISCUSSION THREAD

**Discussion by WeiTangLau** [k3s-io/k3s#2994] [Q&A]
I am planning to deploy a K3s HA cluster with an external DB.

According to the [docs](https://rancher.com/docs/k3s/latest/en/installation/ha/), I will need to have an external DB (e.g. MySQL) to store the cluster information. I would assume this means to install the MySQL server on my host machine, outside of the kubernetes cluster. However, I am planning to use this same database to store my application data. This would mean that the server will not be hosted on kubernetes. (please correct me if I am wrong).

How should I overcome this problem?

Furthermore, I have read the [embedded DB](https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/) option. However, there are very few guides on this (as it recently came out of beta). From my understanding, each of the k3s servers will have its own embedded etcd to store the cluster information. As such, what is the required memory size of the k3s servers (this is crucial as I am running on raspberry pi). Am I able to schedule jobs to the k3s servers (similar to the external db option)? 

PS: I am planning to use 1 of the Raspberry Pis (w/ external SSD) for the PV storage).

I am new to K3s please correct me or ask for further clarification if I did not make myself clear!

Thanks in advance!

**Comment by remkolems** [comment]
The master nodes require either an embedded or an external database to - indeed - store exclusively cluster information.

The worker nodes can host other database(s). Those databases will be and are going to be managed by K3s itself (with correct and proper configuration). 

[Timescaledb](https://www.timescale.com/) (postgresql on steroids) is i.e. an interesting suggestion on my behalf regarding time-series data.

The master nodes with embedded etcd require at least 1GB of memory. Buy the 4GB Raspberry pi 4b w/ an external USB 3 sata SSD drive.

References

1. Take a look @ https://clusberry.techbase.eu/ (not affiliated) for more ideas to create a cluster. 
2. See also https://www.cnx-software.com/2021/01/18/clusberry-9500-cm4-a-raspberry-pi-cm4-cluster-industrial-style/
3. See https://rpi4cluster.com/

**Comment by WeiTangLau** [comment]
@remkolems thanks for replying!

Maybe I should have made myself clear about my configuration. Currently I have 3 Raspberry Pi 4B (1 of which has an external SSD).

Given the 1 SSD limitation, I am planning to adopt the external DB architecture.

Or should I do get 2 more SSDs and use the embedded DB architecture?

**Comment by brandond** [comment]
If you have an external non-Pi server that you are planning on running a database on anyway, I would probably go with that. It'll save a lot of fussing about with SD card IO throughput and such, and keep the SSD free for your PVs.

**Comment by WeiTangLau** [comment]
@brandond sad to say, but I do not have an external non-Pi server available to me (due to limited budget).

Will it be possible, for external DB architecture, use 2 as masters and workers (Ideally should be separated) and the one with SSD will be used as the external DB for cluster info and for PV?

**Comment by brandond** [comment]
Oh, if you don't have an external host then there's not really much point in putting the database on one of the Pis but unmanaged by K3s. I would probably just go with a single server on the node with a SSD, and make the others agents only. I would also boot off the SSD if at all possible.

**Comment by WeiTangLau** [comment]
@bradtopol thanks for your prompt feedback!

Just to clarify. For my setup, it is recommended to adopt the Single-server setup with an Embedded DB architecture with 1 server (Pi w/ SSD) and 2 agents (Pi w/o SSD)?

How do I provision the PV that resides on my server?

Yes currently I am booting from my SD card

**Comment by praveenmak** [comment]
@bradtopol  may I know how I can configure the server nodes to talk to each other using embedded db (etcd). 
If I have 3 server nodes , can I just start k3s server with **--datastore-endpoint etcd** ?

**Comment by brandond** [comment]
I'm not sure who @bradtopol is but the docs for embedded etcd are at https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/

**Reply by AlaaAttya** [reply]
@brandond i've tried to follow all the steps in this document and it fails while validating the certificate and also it's try to connect to etcd using localhost. are you sure that this document is up to date? t https://rancher.com/docs/k3s/latest/en/installation/ha-embedded/

**Reply by brandond** [reply]
Yes, it is.

Are you using TLS on your database server? If so, does it have a valid certificate, or have you disabled ssl certificate validation?

Kine (the etcd to sql translation layer) looks to Kubernetes like an etcd server running on a unix socket, so you will still see logs from Kubernetes about etcd.
